src/go.ts

```diff
@@ -0,0 +1,298 @@
+    }
+
+    this.version = props.version;
+    this.dir = props.dir ?? path.join(process.cwd(), 'dist/go');
+    this.gitBranch = props.branch ?? 'main';
+    this.dryRun = props.dryRun ?? false;
+    this.gitUsername = props.username ?? utils.shell('git config user.name', { capture: true });
+    this.gitUseremail = props.email ?? utils.shell('git config user.email', { capture: true });
+
+    if (this.gitUseremail === '') {
+      throw new Error('Unable to detect username. either configure a global git user.name or pass GIT_USER_NAME env variable');
+    }
+
+    if (this.gitUsername === '') {
+      throw new Error('Unable to detect user email. either configure a global git user.email or pass GIT_USER_EMAIL env variable');
+    }
+
+  }
+
+  /**
+   * Run the release process.
+   *
+   * @returns metadata about the release.
+   */
+  public release(): GoRelease {
+
+    const modules = this.collectModules(this.dir);
+
+    if (modules.length === 0) {
+      console.log('No modules detected. Skipping');
+      return {};
+    }
+
+    console.log('Detected modules:');
+    modules.forEach(m => console.log(` - ${m.modFile}`));
+
+    const repoURL = this.extractRepoURL(modules);
+    const repoDir = path.join(utils.makeTempDirectory(), 'repo');
+    utils.gitHubClone(repoURL, repoDir);
+
+    process.chdir(repoDir);
+
+    try {
+      utils.shell(`git checkout ${this.gitBranch}`);
+    } catch (err) {
+      utils.shell(`git checkout -b ${this.gitBranch}`);
+    }
+
+    this.syncModules(repoDir);
+    try {
+      utils.shell('git add .');
+      utils.shell('git diff-index --exit-code HEAD --');
+      console.log('No changes. Skipping release');
+      return {};
+    } catch (err) {
+      // changes exist, thats ok.
+    }
+
+    const commitMessage = process.env.GIT_COMMIT_MESSAGE ?? this.createReleaseMessage(modules);
+
+    utils.shell(`git config user.name ${this.gitUsername}`);
+    utils.shell(`git config user.email ${this.gitUseremail}`);
+    utils.shell(`git commit -m "${commitMessage}"`);
+
+    const tags = modules.map(m => this.createTag(m));
+
+    if (this.dryRun) {
+      console.log(`Will push to branch: ${this.gitBranch}`);
+      tags.forEach(t => console.log(`Will push tag: ${t}`));
+    } else {
+      utils.shell(`git push origin ${this.gitBranch}`);
+      tags.forEach(t => utils.shell(`git push origin ${t}`));
+    }
+    return { tags, commitMessage };
+  }
+
+  private collectModules(dir: string): GoModule[] {
+    const modules: GoModule[] = [];
+
+    for (const p of [...fs.readdirSync(dir), '.']) {
+      const modFile = path.join(dir, p, 'go.mod');
+      if (fs.existsSync(modFile)) {
+        modules.push(this.createModule(modFile));
+      }
+    }
+    return modules;
+  }
+
+  private createModule(modFile: string): GoModule {
+
+    const version = this.extractVersion(path.dirname(modFile));
+    const majorVersion = version.split('.')[0];
+
+    const cannonicalNameParts = [];
```

Comment:
Add some comments explaining what this array will contain and how the go.mod file is parsed (perhaps a short example)

Reply:
👍